toolchain/musl: update to 1.2.2

neheb opened the pull request:
Remove PKG_SOURCE and PKG_SOURCE_SUBDIR since the defaults are used.

Remove GLOB_ONLYDIR patch. Only fstools relies on it. fstools can be fixed
separately.

Remove woresize.h file. It seems to be for an old version of GCC.

Remove features.h and glibc-types files. Same as above.

Replace sys/cdefs.h with what is in alpine, openembedded, and yocto. The
actual header is deprecated and should not be used. This version is
smaller than the glibc version currently in the tree.

Explanation from buildroot:

  sys/cdefs.h is a non-standard header, originating from glibc,
  that defines non-standard and legacy macros. This header is not
  always available (e.g. musl does not provide it, but Buildroot
  installs a surrogate, minimalist one in this case).

  Some packages use this header, which is wrong, as that makes
  then non-portable.

  By default, Buildroot does not detect the use of that header.
  In the future, this default may change to warning, or even to
  erroring out.

Tested with all packages in the base tree. They all compile.

Signed-off-by: Rosen Penev <rosenp@gmail.com>

The rest of the patches have been sent to the mailing list.

Note that this doesn't seem to work on mips. I noticed upstream has a bunch of patches fixing several things in mips, but not sure which makes mac80211_hwsim panic the kernel on malta.

Posting here as this gets no action on the mailing list.

neheb committed:
toolchain/musl: update to 1.2.2

This release introduces 64-bit time_t, which is needed to avoid the
year 2038 problem.

Remove upstream patches. Refreshed others.

Rebased features.h file based on latest musl.

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
toolchain/musl: remove several GNU headers

Remove GLOB_ONLYDIR patch. Only fstools relies on it. fstools has been
fixed separately.

Remove woresize.h file. It seems to be for an old version of GCC.

Remove features.h and glibc-types files. Same as above.

Remove sys/cdefs.h. This is a deprecated header. Patches to fix packages
that use it have already been patched.

Tested with all packages in the base tree. They all compile.

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
gdb: add openembedded patches

Update to musl 1.2.0 removes sgidefs.h. This causes various compilation
issues. Add patches from openembedded to fix.

Remove uClibc-ng patch. The bug was actually with libstdc++ which was
fixed in 1647790ae6202bd70b199e2aa44f8eb43e8d3af1

Signed-off-by: Rosen Penev <rosenp@gmail.com>

hnyman commented on package/system/fstools/patches/010-glob-onlydir.patch:
```diff
@@ -0,0 +1,158 @@
+Return-Path: <openwrt-devel-bounces+incoming=patchwork.ozlabs.org@lists.openwrt.org>
+X-Original-To: incoming@patchwork.ozlabs.org
+Delivered-To: patchwork-incoming@bilbo.ozlabs.org
+Authentication-Results: ozlabs.org; spf=pass (sender SPF authorized)
```
There is 100+ lines of SMTP protocol chatter, quite unncessary.

hnyman replied:
And there should naturalyl be also a version bump for fstools.

neheb replied:
Replaced with openembedded patch. And done.

neheb commented:
Meh. Those are just placeholders to get them to compile. They should not be merged as is.

hauke commented on package/system/fstools/patches/010-glob-onlydir.patch:
```diff
@@ -0,0 +1,158 @@
+Errors-To: openwrt-devel-bounces+incoming=patchwork.ozlabs.org@lists.openwrt.org
+
+This was originally a patch for musl. It is not present in musl as it is a
+GNU extension.
```
Please extend the description.

GLOB_ONLYDIR is not supported by musl, it is used to restrict the output of glob to directories only, the code has some extra handling to filter out other files for libc implementations that do not support GLOB_ONLYDIR

neheb replied:
Replaced with openembedded patch.

hauke commented on package/utils/xfsprogs/Makefile:
```diff
@@ -8,24 +8,25 @@
 include $(TOPDIR)/rules.mk
```
Please extend the commit description like @aparcar already requested when you send this to the mailing list.

I would prefer to apply this already before the other changes.

neheb replied:
Done

hauke commented on package/system/procd/patches/010-format:
```diff
@@ -0,0 +1,37 @@
+ 	}
+ 
+-	DEBUG(2, "stop %s %s - took %lu.%09lus\n", s->file, s->param, ts_res.tv_sec, ts_res.tv_nsec);
++	DEBUG(2, "stop %s %s - took %" PRId64 ".%09" PRId64 "s\n", s->file, s->param, (int64_t)ts_res.tv_sec, (int64_t)ts_res.tv_nsec);
```
I would prefer "long long" here instead of int64_t Then you can just print it with %llu.

neheb replied:
@dedekeh wanted int64_t for odchpd:\.

hauke commented on package/devel/gdb/patches/140-sgidefs.patch:
```diff
@@ -0,0 +1,11 @@
+ #include "gregset.h"
+ 
+-#include <sgidefs.h>
++#include <asm/sgidefs.h>
```
It would be nice to get this into upstream gdb.

neheb replied:
Doubt they will accept. The GNU sgidefs.h undefines a bunch of stuff from asm/sgidefs.

hauke commented on toolchain/musl/include/bits/wordsize.h:
```diff
@@ -1 +0,0 @@
-#include <sys/user.h>
```
Could you please split this into two commit, one removing these extra headers and one to update musl to 1.2.0.

neheb replied:
Done.

hauke commented:
I like this update of musl to version 1.2, but I expect there some fallout when we merge this.
The changes to fstools and procd should be integrated in the main repository, but having them as patches in this pull request is ok.

neheb commented:
Yes. Do note that this PR is still unable to boot on MIPS. On Malta, it's failing on mac80211_hwsim. After it loads it it seems there's a CPU stall. I've also tried this on my equipment. Similar issue (haven't actually used serial console to find out what the problem is).

After this version release, there were a ton of patches upstream for MIPS. Maybe some need to be backported.

neheb commented:
Marked as ready. It seems mips is working now.

I backported two extra busybox patches as the previous glibc ones caused compilation issues that were also fixed upstream.

adrianschmutzler commented:
So, now the question is whether we want this in and risk regressions/time lost until it is stable everywhere, or wait after a 20.xx branch ...

neheb commented:
Good question.

base gets completely built with this.

packages I'm guessing mostly. I already merged a bunch of patches for that. There are probably several holdouts.

No real idea about the others.

Silent integer truncation still happens in several programs. Not really an issue for the kernel but I assume come 2038 some programs will break. This PR doesn't magically fix that.

20.xx will probably continue to be used past 2038 so there's that.

As far as testing goes, I've only compile tested mips32 and powerpc (both big endian). I ran test mips32 on malta.

mips64, aarch64, arm, and i386 are left. I'll get to those when I can.

neheb commented:
Turns out there are more compile errors after rebasing. Added patches for them.

neheb commented:
Back to draft...

I compile tested on all platforms. Found problems and fixed them.

However it turns out that this is completely broken on big endian devices. The error is in mac80211. Once it loads it causes an RCU stall. I have no idea how to debug this.

ynezz commented:
> So, now the question is whether we want this in and risk regressions/time lost until it is stable everywhere, or wait after a 20.xx branch ...

Indeed, this should be merged right after 20.y is branched, so we've enough time to iron out the remaining issues.

neheb commented:
I sent a bunch of these patches to the mailing list. Let's see what happens...

Grommish commented:
I built the patch series out and ran into no problems on my target - Built/Flashed/Installed

It won't mean too much, as it's an unofficial device, but it doesn't break anything here :+1: 

Build: r13896+12-6269f10ea4 (master branch to commit: 6269f10ea4c4c126a00789e5e0e224422a2b1317)
Arch: mips64/octeon3/gcc10

neheb commented:
Updated the series to build with GCC 10. Although that's not really relevant to musl 1.2.0...

guidosarducci commented:
@neheb   I had been preparing a patch for upstream musl as part of some BPF updates, and also thought about getting the 1.2.0 release for OpenWrt, so this is really great to see!

While testing I couldn't seem to get mips64/malta to even run -- it hung on a boot loop at preinit -- so I wonder if you were able to finally test mips64 as you said above?

Side note: I really wish we would re-enable malta builds for mips64_{be,le} and mips32_le. Without regular builds I'm afraid the code will just accumulate more breakage over time, and it's a shame given how useful it is to have a single platform that covers 64/32-bit and little/big-endian.

neheb commented:
I’ve never gotten Malta under mips(el)64 working. Not even under current musl. Only mips(el)32.

However after this patch set, mips32 gets an RCU stall. The issue is in mac80211. No idea what’s wrong.

neheb commented:
Rebased.

neheb commented:
Rebased and added ppp patch.

guidosarducci commented:
FYI, musl 1.2.1 was just released and includes some mips fixes that might be relevant.

neheb commented:
Still freezes on big endian mips (RCU stall).

guidosarducci commented:
> I’ve never gotten Malta under mips(el)64 working. Not even under current musl. Only mips(el)32.
> 
> However after this patch set, mips32 gets an RCU stall. The issue is in mac80211. No idea what’s wrong.
> 
> Updated and rebased to 1.2.1. I'll test when I can.  Still freezes on big endian mips (RCU stall).
>
A couple of data points... (BTW, I'm trying to get to your configuration step-by-step, but each step reveals other problems.)

First, I figured out the malta/mipsel64 build problems with binutils 2.34 (could you review https://github.com/openwrt/openwrt/pull/3288 please?) and managed to build an image with kernel 5.4 (but old musl 1.1.24). This starts up but still gets a SIGSEGV from procd init and reboots in a loop (see [FS#3277](https://bugs.openwrt.org/index.php?do=details&task_id=3277)).

I thought the SIGSEGV might be related to an earlier memory allocation failure ([FS#3278](https://bugs.openwrt.org/index.php?do=details&task_id=3278)), and found a way to avoid triggering the failure (please review https://github.com/openwrt/openwrt/pull/3289), but the procd init failure and boot loop persists.

I should also note these boot loops don't occur using kernel 4.19 on malta/mips64. (@Grommish Did your build above use kernel 5.4? If so I'd like to see the kernel config as reference...)

I next upgraded to GCC10 (still old musl). This exposed another package build failure that required a "-fno-common" fix in https://github.com/openwrt/packages/pull/13058. My images for malta/mips32 booted OK, but malta/mips64 showed the same procd init failure.

@neheb My takeaways so far:

1. malta/mips64 had problems before using GCC10 and musl 1.2.1, so no point in further musl testing until the underlying mips64 problem is figured out. I could use some help from a procd expert to look at [FS#3277](https://bugs.openwrt.org/index.php?do=details&task_id=3277). Any suggestions?
2. malta/mips32 using GCC10 with kernel 5.4 didn't show RCU problems using musl 1.1.24. That points the finger more strongly at musl 1.2.1.
3. I suggest you try malta/mips32 and malta/mips64 using the older supported kernel 4.19 too. Previously, mips32 didn't show RCU problems, and mips64 had no boot loop problems. This could help focus on a kernel vs. musl problem... 

Next step when I have a chance will be to try musl 1.2.1 with the newer GCC and kernel.

Grommish commented:
@guidosarducci I use 5.4.  But again, I only had to change the musl common.mk itself without the rest of it, so take what you can from it.

Attached is my /proc/config.gz
[config.gz](https://github.com/openwrt/openwrt/files/5047956/config.gz)

neheb commented:
> @guidosarducci I use 5.4. But again, I only had to change the musl common.mk itself without the rest of it, so take what you can from it.
> 
> Attached is my /proc/config.gz
> [config.gz](https://github.com/openwrt/openwrt/files/5047956/config.gz)

The features.h file needs adjusting due to upstream changes between 1.1 and 1.2, Projects that use the header will fail instantly.

neheb commented:
> > I’ve never gotten Malta under mips(el)64 working. Not even under current musl. Only mips(el)32.
> > However after this patch set, mips32 gets an RCU stall. The issue is in mac80211. No idea what’s wrong.
> > Updated and rebased to 1.2.1. I'll test when I can.  Still freezes on big endian mips (RCU stall).
> 
> A couple of data points... (BTW, I'm trying to get to your configuration step-by-step, but each step reveals other problems.)
Want a .config file?
> 
> First, I figured out the malta/mipsel64 build problems with binutils 2.34 (could you review #3288 please?) and managed to build an image with kernel 5.4 (but old musl 1.1.24). This starts up but still gets a SIGSEGV from procd init and reboots in a loop (see [FS#3277](https://bugs.openwrt.org/index.php?do=details&task_id=3277)).
Added patch.
> 
> I thought the SIGSEGV might be related to an earlier memory allocation failure ([FS#3278](https://bugs.openwrt.org/index.php?do=details&task_id=3278)), and found a way to avoid triggering the failure (please review #3289), but the procd init failure and boot loop persists.
> 
> I should also note these boot loops don't occur using kernel 4.19 on malta/mips64. (@Grommish Did your build above use kernel 5.4? If so I'd like to see the kernel config as reference...)
> 
> I next upgraded to GCC10 (still old musl). This exposed another package build failure that required a "-fno-common" fix in [openwrt/packages#13058](https://github.com/openwrt/packages/pull/13058). My images for malta/mips32 booted OK, but malta/mips64 showed the same procd init failure.
> 
> @neheb My takeaways so far:
> 
>     1. malta/mips64 had problems before using GCC10 and musl 1.2.1, so no point in further musl testing until the underlying mips64 problem is figured out. I could use some help from a procd expert to look at [FS#3277](https://bugs.openwrt.org/index.php?do=details&task_id=3277). Any suggestions?
> 
>     2. malta/mips32 using GCC10 with kernel 5.4 didn't show RCU problems using musl 1.1.24. That points the finger more strongly at musl 1.2.1.
I'm testing with GCC9 as GCC10 has some issue with valgrind (I compile with CONFIG_ALL). I think I tried 10 before with the same result.
> 
>     3. I suggest you try malta/mips32 and malta/mips64 using the older supported kernel 4.19 too. Previously, mips32 didn't show RCU problems, and mips64 had no boot loop problems. This could help focus on a kernel vs. musl problem...
> 
> 
> Next step when I have a chance will be to try musl 1.2.1 with the newer GCC and kernel.

neheb commented:
Rebased.

I tried reverting toolchain/musl: remove several GNU headers . Didn't do anything. mips32 still fails.

Grommish commented:
@guidosarducci I should also mention that I'm running gcc10 along with the musl 1.2.0 (i've not tried 1.2.1 yet)

neheb commented:
Rebased and reworked the ppp patch.

Removed sys/cdefs. It seems it's not really needed. I have a bunch of patches for the packages feed that fix sys/cdefs.h usage.

guidosarducci commented:
> > > I’ve never gotten Malta under mips(el)64 working. Not even under current musl. Only mips(el)32.
> > > However after this patch set, mips32 gets an RCU stall. The issue is in mac80211. No idea what’s wrong.
> > > Updated and rebased to 1.2.1. I'll test when I can.  Still freezes on big endian mips (RCU stall).
> > 
> > A couple of data points... (BTW, I'm trying to get to your configuration step-by-step, but each step reveals other problems.)
> 
> Want a .config file?
> 
Ah, no, I meant I was changing things systematically, one-at-a-time, to try identifying where problems arose i.e. proper A/B testing.
> > 
> > Next step when I have a chance will be to try musl 1.2.1 with the newer GCC and kernel.
> > 
I pulled the trigger and tested musl 1.2.1 along with kernel 5.4 and GCC10, together with your patches, with the expected results: malta-be hits the RCU stall after a 60-second timeout, while malta-be64 hits the procd init failure and reboot before reaching the RCU stall point.

What's strange about the RCU stall is why/how a userspace library change would lead to a kernel hang at early boot. The stall also happens on kernel 4.19, so certainly seems a musl issue. Comparing a malta-be boot log using musl 1.1.24, this event sequence would normally happen instead of the RCU stall:
```
 mac80211_hwsim: initializing netlink
 ieee80211 phy0: Selected rate control algorithm 'minstrel_ht'
 ieee80211 phy1: Selected rate control algorithm 'minstrel_ht'
[...]
```
Smoking gun seems to be the `mac80211_hwsim` wifi simulator, which uses netlink to talk to userspace. One guess is that musl 1.2.1 broke something (perhaps libnl??) on the netlink userspace side. On malta platform `kmod-mac80211-hwsim` is a default package, and I discovered if you disable this in the config then the system will boot properly without the RCU stall.

So we add `kmod-mac80211-hwsim` to the NEEDS_FIXING list, but at least other testing using malta can continue. Hope that helps.

Grommish commented:
> Smoking gun seems to be the `mac80211_hwsim` wifi simulator,

This would make sense, since the itus device doesn't have Wifi.  it would explain why I am not seeing any of the issues everyone else is.

neheb commented:
Not quite.

mac80211 breaks with real hardware too. Not just hwsim.

Didn't know libnl could be the issue. I'll look into it when I can.

neheb commented:
https://git.kernel.org/pub/scm/linux/kernel/git/next/linux-next.git/commit/?h=next-20200810&id=551599edbfff2431cef943a772fbde1c3e26eaf8

shows CONFIG_COMPAT_32BIT_TIME . I removed that but it still seems that it fails.

neheb commented:
I made a new branch, isolating most of the important changes to try to get to the bottom of this...

https://github.com/neheb/openwrt/tree/musl120-test

Upstream says the kernel does not use any musl code.

HOWEVER, OpenWrt uses the backports tree. It does not use the stuff built into the kernel. Which makes me believe that it's being built with musl...

One other hint is that I have a small mac80211 change to fix compilation. Which would be weird unless musl was being used. I mean even the description indicates so:

```
This define is not only useless, it breaks the build as the little
endian header always gets included, which subsequently errors.
```

I have a feeling the backports tree is always building little endian code, even for big endian devices.

guidosarducci commented:
FYI, I just pushed a fix for the malta/mips64 crash at init and bootloop: https://github.com/openwrt/openwrt/pull/3307.  You could try this to see now if the mac80211 problem is word-size or endian sensitive.

neheb commented:
> FYI, I just pushed a fix for the malta/mips64 crash at init and bootloop: #3307. You could try this to see now if the mac80211 problem is word-size or endian sensitive.

Endian problem. malta le64 works fine. so does le. be and be64 fail.

neheb commented:
Rebased.

Added util-linux update. Turns out, util-linux' hwclock has been totally broken with musl when using the -k parameter. Upstream fixed that. Also backported upstream's musl 1.2.x fix.

neheb commented:
Also turns out that busybox hwclock has been broken in the same way. Added patch to fix that as well.

neheb commented:
Reworked the busybox fix.

Removed the date -k patch and replaced with hwclock --systz. The code is mostly identical anyway.

Backported two patches to that end.

neheb commented:
While looking at unrelated stuff, I came across this: https://git.musl-libc.org/cgit/musl/commit/?id=97d35a552ec5b6ddf7923dd2f9a8eb973526acea

I have a feeling this is breaking mac80211 on big endian as now __LITTLE_ENDIAN and __BIG_ENDIAN are defined unconditionally. I'm guessing there's some ifdef __LITTLE_ENDIAN codepath breaking things.

guidosarducci commented:
Interesting. Do you have any new insights into how userspace defs might leak into mac80211 code in the first place? I didn't spot anything in the limited poking around I could do. I did find that some registers/structures used for 80211 are fixed endianness though, so there's some built-in awareness of that. And nothing relevant jumped out while looking through upstream mac80211 commits either.

I think debugging the startup of the `mac80211_hwsim` module could yield some better clues. Will try that if I can find some time.

neheb commented:
> Interesting. Do you have any new insights into how userspace defs might leak into mac80211 code in the first place? I didn't spot anything in the limited poking around I could do. I did find that some registers/structures used for 80211 are fixed endianness though, so there's some built-in awareness of that. And nothing relevant jumped out while looking through upstream mac80211 commits either.
> 
> I think debugging the startup of the `mac80211_hwsim` module could yield some better clues. Will try that if I can find some time.

I see includes to stdio.h and whatnot. That's going to pull in alltypes.h from musl. Prior to that commit, endian.h needed to be included for definitions. One annoying thing is that GCC has no option to error on macro redefinition. It doesn't emit diagnostics for headers at all. clang does but not GCC.

neheb commented:
Removed busybox update. I guess there's hesitation on updating it. Backported three patches instead.

neheb commented:
Rebased.

neheb commented:
@hauke managed to figure it out. musl's headers were indeed leaking into backports. Update pushed.

neheb commented on package/libs/libroxml/Makefile:
```diff
@@ -9,36 +9,36 @@ include $(TOPDIR)/rules.mk
-PKG_LICENSE:=LGPL-2.1+
 
 include $(INCLUDE_DIR)/package.mk
+include $(INCLUDE_DIR)/cmake.mk
```
IIRC no.

neheb commented on package/network/utils/iftop/patches/010-gcc10.patch:
```diff
@@ -0,0 +1,20 @@
+-hash_type* screen_hash;
+-hash_type* service_hash;
++static hash_type* screen_hash;
++static hash_type* service_hash;
```
I agree. I'll rebase and resend.

neheb commented:
Rebased.  Fixed trace-cmd picking up libaudit. Removed upstreamed stuff.

neheb commented:
Rebased again...

neheb commented:
Rebased.

@guidosarducci I noticed yet another problem with bpftools + musl 1.2.x + powerpc. __always_inlined ends up being undefined.

neheb commented:
Rebased again. This is getting quite tiring...

neheb committed:
bpftools: fix compilation with musl 1.2.x

A definition for __maybe_inline is needed.

Refreshed patches.

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
umbim: do not warn for address of pointer member

Warning was introduced in GCC9.

Signed-off-by: Rosen Penev <rosenp@gmail.com>

guidosarducci commented:
> Rebased.
> 
> @guidosarducci I noticed yet another problem with bpftools + musl 1.2.x + powerpc. __always_inlined ends up being undefined.

Still catching up after a while, so thanks for the heads up. Do you have a some compiler logs, or thoughts on root cause? I'm trying to understand what if anything should be up-streamed and to where... Stay safe and best in the coming year.

neheb commented:
```
--- a/tools/bpf/bpftool/btf_dumper.c
+++ b/tools/bpf/bpftool/btf_dumper.c
@@ -5,6 +5,7 @@
 #include <stdio.h> /* for (FILE *) used by json_writer */
 #include <string.h>
 #include <unistd.h>
+#include <linux/compiler.h>
 #include <asm/byteorder.h>
 #include <linux/bitops.h>
 #include <linux/btf.h>
--- a/tools/bpf/bpftool/map_perf_ring.c
+++ b/tools/bpf/bpftool/map_perf_ring.c
@@ -16,6 +16,7 @@
 #include <time.h>
 #include <unistd.h>
 #include <linux/bpf.h>
+#include <linux/compiler.h>
 #include <linux/perf_event.h>
 #include <sys/ioctl.h>
 #include <sys/mman.h>
```

is the patch. I think a definition is needed by one of the headers.

hnyman commented:
Like I commented in the busybox upgrade discussion the proposed busybox change in this PR from "date -k" to "hwclock --systz" leads into strange timestamps on bootlogs, at least on devices that do not have a real real-time clock.

http://lists.openwrt.org/pipermail/openwrt-devel/2021-January/033240.html
http://lists.openwrt.org/pipermail/openwrt-devel/2021-January/033243.html

neheb commented:
Can you try adding -u to that hwclock call?

Looking at the hwclock and 250-date-k.patch , the code is basically the same except for:

```
+		memset(&tz, 0, sizeof(tz));
+		syscall(SYS_settimeofday, NULL, &tz);
+
+		memset(&tz, 0, sizeof(tz));
```

The hwclock code does an extra settimeofday call if -u gets passed.

guidosarducci commented:
> ```
> --- a/tools/bpf/bpftool/map_perf_ring.c
> +++ b/tools/bpf/bpftool/map_perf_ring.c
> @@ -16,6 +16,7 @@
>  #include <time.h>
>  #include <unistd.h>
>  #include <linux/bpf.h>
> +#include <linux/compiler.h>
>  #include <linux/perf_event.h>
>  #include <sys/ioctl.h>
>  #include <sys/mman.h>
> ```
> 
> is the patch. I think a definition is needed by one of the headers.

Ah no, I've already seen the patch but was looking to understand the underlying problem better -- hence asking for compiler logs, etc.  Otherwise I'll need to find time to set up the build environment and try to reproduce...

neheb commented:
Updated to musl 1.2.2, kexec-tools 2.0.21, and busybox 1.33.

hnyman commented:
> Can you try adding -u to that hwclock call?

Seems to work. Timestamps on the system log look sensible both after  flash and after a normal reboot.

I assume that it has also worked for you in your own testing with live routers?


```
--- a/package/base-files/files/etc/init.d/system
+++ b/package/base-files/files/etc/init.d/system
@@ -27,7 +27,7 @@ system_config() {
 		ln -sf "/usr/share/zoneinfo/$zonename" /tmp/localtime && rm -f /tmp/TZ
 
 	# apply timezone to kernel
-	busybox date -k
+	hwclock -u --systz
 }
 
 reload_service() {
```

neheb commented:
> > Can you try adding -u to that hwclock call?
> 
> Seems to work. Timestamps on the system log look sensible both after flash and after a normal reboot.
> 
> I assume that it has also worked for you in your own testing with live routers?
It has been a while since I've flashed any routers... The only device I have deployed is a Turris Omnia with TurrisOS on it. Although it seems that I will be retiring it again...

I've tried using malta but it seems there's an RTC there. I also tried running it locally on Fedora 33 but something weird happens there. Maybe a difference between util-linux hwcloick and busybox hwclock.
> 
> ```
> --- a/package/base-files/files/etc/init.d/system
> +++ b/package/base-files/files/etc/init.d/system
> @@ -27,7 +27,7 @@ system_config() {
>  		ln -sf "/usr/share/zoneinfo/$zonename" /tmp/localtime && rm -f /tmp/TZ
>  
>  	# apply timezone to kernel
> -	busybox date -k
> +	hwclock -u --systz
>  }
>  
>  reload_service() {
> ```

hnyman commented:
> It has been a while since I've flashed any routers... 

That sounds risky. 
You are trying to upgrade the deepest core components / building blocks of OpenWrt, meaning musl (all C support) and busybox (all shell support). If you really mean that you have not even tested the proposed changes in real-life by yourself in several different routers, that sounds really scary to me.

I know that you are doing lots of work, but tinkering with these deeply intertwined core components is a bit different that a run-of-the-mill upgrade of a singular vanilla userspace package.

neheb commented:
Let me correct that. I've ran this update on my Archer C7v2 with 1.2.1 and busybox 1.3.2 with no real issues. As you alluded to, I never paid attention to log timestamps during early boot. My earlier comments had mainly to do with that.

The small C file in the hwclock commit was ran on actual hardware (cross-compiled).

SquallATF commented:
musl 1.2.2 add gettid function, will cause libselinux package build failed.

https://github.com/SELinuxProject/selinux/blob/7df27b78e9eecbe65a57cdfefb9e51f547231b20/libselinux/src/procattr.c#L25-L42

neheb commented:
this is correct. I was hit by this. Unfortunately there's no good solution...

LynxChaus commented:
JFYI: Under musl 1.2.x external kernel modules (mt76 in my case) for any LITTLE_ENDIAN arch is miscompiled.
Fast & dirty fix - remove include/stdarg.h after install.

neheb commented:
Any? Hauke fixed the issue for mc80211: https://git.kernel.org/pub/scm/linux/kernel/git/backports/backports.git/commit/?id=473fbe303f36b84c6653dd84ee302d1b70078479

I assume a similar one would be needed where libc headers are being leaked.

LynxChaus commented:
@neheb Better remove stdarg.h from installed musl. stdarg.h only contains defines to _builtin_* functions, and gcc carry stdarg.h with himself.

neheb commented:
Why would I edit musl itself?

neheb commented:
Now that 21.02 has been branched, can the musl update be merged?

SquallATF commented:
> 
> 
> this is correct. I was hit by this. Unfortunately there's no good solution...

selinux upstream has fixed this issue https://github.com/SELinuxProject/selinux/commit/398d2ceef92cb1baac18e6b34a1a8e1bf41296cd

neheb commented:
cool.

rsalvaterra commented:
Hm…
```
procattr.c:38:14: error: static declaration of 'gettid' follows non-static declaration
   38 | static pid_t gettid(void)
      |              ^~~~~~
In file included from /home/rui/Sources/OpenWrt/openwrt/staging_dir/toolchain-mips_74kc_gcc-10.2.0_musl/include/fortify/unistd.h:22,
                 from procattr.c:2:
/home/rui/Sources/OpenWrt/openwrt/staging_dir/toolchain-mips_74kc_gcc-10.2.0_musl/include/unistd.h:194:7: note: previous declaration of 'gettid' was here
  194 | pid_t gettid(void);
      |       ^~~~~~
make[5]: *** [Makefile:159: procattr.o] Error 1
```
… I thought the pull request still included the workaround, while libselinux isn't updated. 😕

neheb commented:
Right. I have it patched locally but not did add to this PR.

rsalvaterra commented:
Why? It should actually be the first patch of the series, in order not to break the build, no?

neheb commented:
This PR used to have double the amount of patches. Half of them were merged already.

Btw, the libselinux failure is recent. It didn't always fail to compile.

neheb commented:
kexec-tools were merged. gdb and bpftools commit are all that are left. umdns and umbim are GCC10 specfific.

rsalvaterra commented:
Is there anything else needed, now that the (IMHO) big blocker is [fixed](https://git.openwrt.org/?p=openwrt/openwrt.git;a=commitdiff;h=9ac47ee46918c45b91f4e4d1fa76b1e26b9d57fe)? I've tested this on ath79, mvebu/cortexa9 and ramips/mt7621, and it's now working perfectly. And the higher efficiency of mallocng is in itself a killer feature… 🙂

neheb commented:
Nope. All packages compile.

guidosarducci commented:
> > Rebased.
> > @guidosarducci I noticed yet another problem with bpftools + musl 1.2.x + powerpc. __always_inlined ends up being undefined.
> 
> Still catching up after a while, so thanks for the heads up. Do you have a some compiler logs, or thoughts on root cause? I'm trying to understand what if anything should be up-streamed and to where... Stay safe and best in the coming year.

Following up, could you please fix the commit message for bpftools? It's misleading and doesn't include details of the error. Does the patch still build OK with default settings i.e. musl 1.1.x, GCC 8.4. Any idea what changed in musl to cause the compile failure?

I'd like to minimize our bpf patches, and everything so far has been upstreamed. Could you submit this one as well? Thanks!

neheb commented:
I'd rather not deal with kernel.org

musl 1.2.x has a lot of deep header changes.

The issue is probably the removal of those GNU headers. I doubt bpftools gets built without glibc. Alpine Linux doesn't seem to have a package for it. They would run into the same issue.

SkewedZeppelin commented:
I've had this running on my shelby and caiman for 6 days now with no issues or noticeable regressions.
And no reported issues from any users of my builds.

neheb commented:
Forget it.

rsalvaterra commented:
> Forget it.

WTF, mate…? This was basically ready, just needed to be merged! 😞

neheb commented:
@rsalvaterra feel free to resubmit.

adschm commented:
Since 21.02 has been released by now, and waiting longer won't help make updating any easier, I would like to revive this submission.

@neheb It would be nice if you had a look at whether we need any updates here.

@rsalvaterra Since you have been involved with the subject as far as I can see, I'd be happy if you could review and re-test this another time.

Though I'm definitely leaving my comfort zone here (and have very limited expertise on the details), I think now would be a good window to merge this. Hence, I would take it if we don't see any new complications or build/test issues (unless @rsalvaterra  wants to take the blame and use his recently gained commit rights :-) )

neheb commented:
@hauke already posted this on the mailing list. I’d just wait for him.

rsalvaterra commented:
@adschm, I don't "test" musl 1.2.2 anymore, it's just part of my tree and all my systems are running it. :) As far as I can tell, it's 100 % stable and the higher allocator efficiency (less memory footprint) is the cherry on top.

adschm commented:
> @hauke already posted this on the mailing list. I’d just wait for him.

Indeed, must have overlooked that. So probably I should just give it an Ack there.

hnyman commented:
I noticed problems with collectd rrd database (my stats history for the lat 12 months), when I restored it in the new musl 1.2.2 enabled master build. I had saved the stats a few minutes earlier with yesterday's build (with 1.1.24).

collectd complains 

```
Thu Sep 23 19:32:27 2021 daemon.err collectd[7290]: rrdtool plugin: rrd_update_r failed: /tmp/rrd/router1/cpu-1/percent-softirq.rrd: illegal attempt to update using time 1313558100 when last update time is 1632414748 (minimum one second step)
Thu Sep 23 19:32:27 2021 daemon.err collectd[7290]: rrdtool plugin: rrd_update_r failed: /tmp/rrd/router1/cpu-1/percent-idle.rrd: illegal attempt to update using time 1313558100 when last update time is 1632414748 (minimum one second step)
Thu Sep 23 19:32:27 2021 daemon.err collectd[7290]: rrdtool plugin: rrd_update_r failed: /tmp/rrd/router1/load/load.rrd: illegal attempt to update using time 1313558100 when last update time is 1632414748 (minimum one second step)
Thu Sep 23 19:32:27 2021 daemon.err collectd[7290]: rrdtool plugin: rrd_update_r failed: /tmp/rrd/router1/interface-br-lan/if_packets.rrd: illegal attempt to update using time 959657776 when last update time is 1632414748 (minimum one second step)
Thu Sep 23 19:32:27 2021 daemon.err collectd[7290]: rrdtool plugin: rrd_update_r failed: /tmp/rrd/router1/interface-br-lan/if_octets.rrd: illegal attempt to update using time 808989488 when last update time is 1632414748 (minimum one second step)
Thu Sep 23 19:32:27 2021 daemon.err collectd[7290]: rrdtool plugin: rrd_update_r failed: /tmp/rrd/router1/interface-br-lan/if_errors.rrd: illegal attempt to update using time 47 when last update time is 1632414748 (minimum one second step)
Thu Sep 23 19:32:27 2021 daemon.err collectd[7290]: rrdtool plugin: rrd_update_r failed: /tmp/rrd/router1/interface-br-lan/if_dropped.rrd: illegal attempt to update using time 47 when last update time is 1632414748 (minimum one second step)
Thu Sep 23 19:32:27 2021 daemon.err collectd[7290]: rrdtool plugin: rrd_update_r failed: /tmp/rrd/router1/cpufreq-0/cpufreq.rrd: illegal attempt to update using time 1313558100 when last update time is 1632414748 (minimum one second step)
Thu Sep 23 19:32:27 2021 daemon.err collectd[7290]: rrdtool plugin: rrd_update_r failed: /tmp/rrd/router1/cpufreq-0/transitions.rrd: illegal attempt to update using time 942945072 when last update time is 1632414748 (minimum one second step)
Thu Sep 23 19:32:27 2021 daemon.err collectd[7290]: rrdtool plugin: rrd_update_r failed: /tmp/rrd/router1/memory/memory-buffered.rrd: illegal attempt to update using time 1313558100 when last update time is 1632414748 (minimum one second step)
Thu Sep 23 19:32:27 2021 daemon.err collectd[7290]: rrdtool plugin: rrd_update_r failed: /tmp/rrd/router1/memory/memory-used.rrd: illegal attempt to update using time 1313558100 when last update time is 1632414748 (minimum one second step)
...
Thu Sep 23 19:32:27 2021 daemon.err collectd[7290]: rrdtool plugin: rrd_update_r failed: /tmp/rrd/router1/cpufreq-1/cpufreq.rrd: illegal attempt to update using time 1313558100 when last update time is 1632414748 (minimum one second step)
Thu Sep 23 19:32:27 2021 daemon.err collectd[7290]: rrdtool plugin: rrd_update_r failed: /tmp/rrd/router1/cpufreq-1/transitions.rrd: illegal attempt to update using time 959459376 when last update time is 1632414748 (minimum one second step)
Thu Sep 23 19:32:27 2021 daemon.err collectd[7290]: rrdtool plugin: rrd_update_r failed: /tmp/rrd/router1/cpufreq-1/percent-384000.rrd: illegal attempt to update using time 1313558100 when last update time is 1632414748 (minimum one second step)
```

There is a difference of roughly 318,856,648 between the archived old last timestamp and the new one, which difference does not immediately ring a bell. Curiously, a few of the plugins show a different current time, (and in subsequent attempts, the same plugins always show the same erroneous timestamp like  808989488 or 47. I guess that some plugin timestamps are not yet compatible with the 64bit time_t.

More strangely, later I noticed that the "rrdtool" shows segmentation error when I tried to look into the contents of the new data stored in a new rrd database. LuCI still shows ok graphs, so "rrdgraph" works.  Might be that our rrdtool handles some values as 32bit and there is overflow.

I haven't yet debugged this further, but strange errors.  (My whole toolchain has been rebuilt today, so there shouldn't be any .)

cc @neheb @jow-

neheb commented:
Oh yeah. A lot of packages need either lld or PRIu64. Probably the latter.

hnyman commented:
Looks like rrdtool has been later updated to handle 64 bit time_t even on 32bit systems, but as we use a really ancient 1.0.50 from year 2005 due to its small footprint (without external graph libs), we have not got the newer fixes.

I haven't yet looked into collectd.

hnyman commented:
I wrote a PR for rrdtool1, but started thinking that I haven't tested it with glibc.
https://github.com/openwrt/packages/pull/16703

Any ideas about that glibc compatiblity angle, @neheb ?

In general:
If we start fixing apps for 64bit time_t in musl, how will glibc users react?
* Do we possibly need additional "(unsigned long long)" casting ? (If glibc still uses 32bit time_t)
* Or can we force/configure glibc to also use 64bit time_t ?
  ( https://www.gnu.org/software/libc/manual/html_node/64_002dbit-time-symbol-handling.html )

neheb commented:
There's only one platform that uses glibc by default (ARC) which has no real devices. I wouldn't worry too much about it. Although since ARC is relatively new to glibc, it's 64-bit by default even though it's 32-bit.

64-bit by default sounds great. I have no idea how to flip the switch. AFAIK, glibc in OpenWrt is a development thing.

hnyman commented:
Regarding toolchain glibc and 64bit time_t:

If we now fix all our packages for the 64bit time_t due to musl 1.2.2 forcing it , it would be great to also configure glibc to always use 64but time_t.  Otherwise the glibc users with 32bit targets will run into trouble (when packages expects 64bit time_t but glibc defaults to 32bit.)

Apparently glibc 2.34 (released last month) now supports forcing 64bit time_t.

https://sourceware.org/pipermail/libc-alpha/2021-August/129718.html
https://sourceware.org/git/?p=glibc.git;a=tag;h=9df03063320651bc629fa427eef3ac73fabb61ba

> * Add support for 64-bit time_t on configurations like x86 where time_t
>   is traditionally 32-bit.  Although time_t still defaults to 32-bit on
>   these configurations, this default may change in future versions.
>   This is enabled with the _TIME_BITS preprocessor macro set to 64 and is
>   only supported when LFS (_FILE_OFFSET_BITS=64) is also enabled.  It is
>   only enabled for Linux and the full support requires a minimum kernel
>   version of 5.1.


@dedeckeh 
Might you look into this, as you have been regularly upgrading glibc in the toolchain...
